### ISSUE TITLE
About GPU utilization

### ISSUE THREAD

**Issue by trunglebka** [rhasspy/piper#121]
This isn't an issue but a question about GPU utilization during training 
I'm training my voice with piper but it does not utilize GPU well. about 30% - I think

Currently I'm running/training two different experiment (not DDP) and this is my screenshot of nvtop
![image](https://github.com/rhasspy/piper/assets/16150250/1a625f34-3df5-4dce-9eee-43c836287b48)

I wonder if there is a method to enhance the situation, or if this is a specific pattern of the model?
My machine has 210 GB free memory that has been used to cache and dataset cache size is jus 15GB (as show in htop) so I don't think that it is problem of slowly disk.

**Comment by Lauler** [comment]
What worked for me: 

The `num_workers` param/arg was not configurable in the Lightning config. In my case training by default only ran with 1 CPU for dataloading. This caused the CPU to bottleneck the training.

Try manually editing `num_workers` arg in the functions `train_dataloader` and `valid_dataloader` in `piper/src/python/lightning.py` and see if that helps.

**Comment by trunglebka** [comment]
I've changed that number (to five workers) but from htop addition threads almost does not use CPU at all while main process keep taking full CPU. GPU usage almost the same as original code without dataloader worker

**Comment by Lauler** [comment]
I see the VRAM used is about 9GB. What size model are you training and with what batch size? If your batch size is small, then that also tends to decrease GPU utilization. 

In my case, when I train a high quality model with batch size 8, the utilization drops to around 50%. But when I train the same model with batch size 32, it averages 80% to 90%.

**Comment by trunglebka** [comment]
@Lauler
I'm training with batch size 8, medium model. I can just run with that batch size because of some initial iteration it has some peak and cause OOM error.
Based on your info I'm suspecting that I'm simply does not have enough VRAM to utilize full of GPU

**Comment by trunglebka** [comment]
Well, look like my ancient CPU not powerful enough to saturate GPU.
[synesthesiam](https://github.com/synesthesiam), can you suggest me some CPU intensive parts that can be cached to reduce CPU usage?

**Comment by synesthesiam** [comment]
I believe the most CPU intense part is the batch collating, where it sorts, pads, and copies all the tensors, which can't really be cached. Piper preprocessing does all the audio conversion already.

Turning off validation and testing phases may help by setting the "--num-...." flags to 0.

**Comment by trunglebka** [comment]
@synesthesiam
I've monitored UtteranceCollate, it run pretty fast, about 0.015s for a batch of size 16.
Using a cache to ensure that UtteranceCollate is not compute expensive, main training process still saturate a CPU.

**Comment by trunglebka** [comment]
Thank you guys, maybe I should try simplest solution: jump ship 😂

**Comment by trunglebka** [comment]
@synesthesiam there is some weird degrading GPU utilization between start of training process and after a "long" time as you can see in the picture below. I don't know if it is the problem with piper or pytorch lightning

![image](https://github.com/rhasspy/piper/assets/16150250/aee38d2d-a98b-445d-a068-a8e7529c1bda)

**Comment by coffeecodeconverter** [comment]
for me, im on windows WSL and the num_workers warning 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
rank_zero_warn(
/home/PiperTTS/.venv/lib/python3.10/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:224: 
PossibleUserWarning: 
The dataloader, train_dataloader, does not have many workers which may be a bottleneck. 
Consider increasing the value of the `num_workers` argument` (try 16 which is the number of cpus on this machine) 
in the `DataLoader` init to improve performance.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

was resolved by editing the lightning.py found here:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/PiperTTS/src/python/piper_train/vits/lightning.py
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

within this class:
~~~~~~~~~~~~~~~~~~~~~
class VitsModel(pl.LightningModule):
    def __init__(
        self,
        num_symbols: int,
        num_speakers: int,
        # audio
        resblock="2",
        resblock_kernel_sizes=(3, 5, 7),
        resblock_dilation_sizes=(
            (1, 2),
            (2, 6),
            (3, 12),
        ),
        upsample_rates=(8, 8, 4),
        upsample_initial_channel=256,
        upsample_kernel_sizes=(16, 16, 8),
        # mel
        filter_length: int = 1024,
        hop_length: int = 256,
        win_length: int = 1024,
        mel_channels: int = 80,
        sample_rate: int = 22050,
        sample_bytes: int = 2,
        channels: int = 1,
        mel_fmin: float = 0.0,
        mel_fmax: Optional[float] = None,
        # model
        inter_channels: int = 192,
        hidden_channels: int = 192,
        filter_channels: int = 768,
        n_heads: int = 2,
        n_layers: int = 6,
        kernel_size: int = 3,
        p_dropout: float = 0.1,
        n_layers_q: int = 3,
        use_spectral_norm: bool = False,
        gin_channels: int = 0,
        use_sdp: bool = True,
        segment_size: int = 8192,
        # training
        dataset: Optional[List[Union[str, Path]]] = None,
        learning_rate: float = 2e-4,
        betas: Tuple[float, float] = (0.8, 0.99),
        eps: float = 1e-9,
        batch_size: int = 1,
        lr_decay: float = 0.999875,
        init_lr_ratio: float = 1.0,
        warmup_epochs: int = 0,
        c_mel: int = 45,
        c_kl: float = 1.0,
        grad_clip: Optional[float] = None,
        num_workers: int = 1,
        seed: int = 1234,
        num_test_examples: int = 5,
        validation_split: float = 0.1,
        max_phoneme_ids: Optional[int] = None,
        **kwargs,
    ):
~~~~~~~~~~~~~~~~~~~~~

changed this default
~~~~~~~~~~~~~~~~~~
num_workers: int = 1
~~~~~~~~~~~~~~~~~~

to anything above 2, otherwise it still gives you the warning
i have 16 cores available, but as an initial test, only increased it to 4 
~~~~~~~~~~~~~~~~~~
num_workers: int = 4
~~~~~~~~~~~~~~~~~~

and i no longer see the warning during training.

**Comment by kevkid** [comment]
Weird, I am also seeing this issue. Seems like my gpu hits 100% util then drops to 30% then jumps to 100% and drops again. My CPU is not saturated. I have changed the num_workers to 10 and each time I run it, I print the num of workers, but still getting these spikes of performance. Any ideas?
